lib/galaxy/jobs/runners/local.py

```diff
@@ -214,10 +214,16 @@ def _check_pid(self, pid):
             return False
 
     def _terminate(self, proc):
-        os.killpg(proc.pid, 15)
-        sleep(1)
-        if proc.poll() is None:
-            os.killpg(proc.pid, 9)
+        process_exists = True
+        try:
+            os.killpg(proc.pid, 15)
```

Comment:
```suggestion
            proc.terminate()
```

Reply:
`os.killpg()` != `Popen.terminate()` right? From [what I can tell](https://docs.python.org/2/library/subprocess.html#subprocess.Popen.terminate), `Popen.terminate()` is not going to signal all pids in the process group?

Reply:
But this is invoking `os.killpg()` on a process id ( https://docs.python.org/2/library/subprocess.html#subprocess.Popen.pid ), not a process group id, right?

Reply:
The call to `subprocess.Popen()` uses `os.setpgrp()` to set the process's pgid to its pid (i.e. ensures that it is the root of its own process group instead of Galaxy's process group), so the pgid to kill is the pid.

Reply:
Ah, TIL! I'll add a comment to document that.

Reply:
@natefoo I did some testing on this today and I noticed that `os.killpg()` was not killing orphan processes left by `proc`. This can be tested with `'python  forking_orphan_processes.py'`  as `command_line` , where ` forking_orphan_processes.py` is https://gist.github.com/nsoranzo/d68793e9fb6269f997eb37d2747c262e

This seems to be fixed by using `preexec_fn=os.setsid` (instead of `preexec_fn=os.setpgrp`) when calling `subprocess.Popen()` . What do you think?

Reply:
Yeah, I can't reproduce it myself, I probably mixed things up with my unsuccessful attempt to use `os.waitpid(-proc.pid, 0)` to wait for the orphan processes. I like the idea of the `os.killpg()` call after `proc.wait()`, I'll work on a patch.